### ISSUE TITLE
Include terms in the owl ontology

### ISSUE THREAD

**Issue by fanavarro** [IHTSDO/snomed-owl-toolkit#20]
Hello, I'm trying to generate an ontology from SnomedCT International edition, the Spanish release and the Spanish extension from the Spanish Ministry. I'm using the following command:
`java -jar snomed-owl-toolkit.jar -rf2-snapshot-archives SnomedCT_International_Edition.zip,SnomedCT_Spanish_Edition.zip,SnomedCT_Extension_Española_SNS.zip`
This command generates an owl file, but it only includes labels (rdfs:label) in English that usually have the form "class label (super class label)", for example "Antichistosomal (disposition)".
I am wondering if it is possible to add preferred and synonyms terms to the ontology as I think this information exists in SnomedCT. Maybe the issue #14 is related with this one.

Thanks in advance,
Francisco Abad.

**Comment by kaicode** [comment]
Hi @fanavarro, you are right #14 is the outstanding feature which would add translated preferred synonyms to the ontology. I'll try finish that feature and get it committed over the next few days. Watch this space.

**Comment by fanavarro** [comment]
I've just seen a new snapshot for the feature, great!
I'll try it in a while.
Thanks!

**Comment by fanavarro** [comment]
I've tried the preview release and it works well, I'm able to generate the ontology with the labels in the specified language reference set. Nonetheless I would like to suggest the possibility of generating the ontology with all the terms by creating different annotations for each extension. For example, the annotation 'es-ES.preferred' for Spanish (Spain) preferred terms or 'es.preferred' for International Spanish preferred terms, or 'en.acceptable' for International English synonyms.

The resulting ontology will show something like the following figure:
![imagen](https://user-images.githubusercontent.com/11376639/51168472-ac412180-18a9-11e9-8f05-405dc947190d.png)

I think this functionality should be taken into account in the future since it could be very useful for NLP tasks, for example.

Thanks,
Francisco Abad.

**Comment by kaicode** [comment]
Thanks for trying the preview release, I'm glad it worked for you.

Your suggestion for a further enhancement including all terms sounds useful and thanks for the screenshot.
Could you provide a snippet of OWL functional syntax to show what that would look like? I couldn't find any documentation on how to mark the language of an annotation in OWL.
It would be useful to have a ticket with all these details so that someone could pick this up. This is an open source project and we welcome code contributions.

**Comment by fanavarro** [comment]
Hi @kaicode, thanks for your answer.
You can specify the language tag for a string by adding "@langCode" to that string. The owl syntax for my example would be the following:
```

Declaration(Class(<http://snomed.info/id/190787008>))
# Class: <http://snomed.info/id/190787008> (Abetalipoproteinemia (disorder))
AnnotationAssertion(<http://snomed.info/field/es-ES.acceptable> <http://snomed.info/id/190787008> "enfermedad de Bassen - Kornzweig"@es-es)
AnnotationAssertion(<http://snomed.info/field/es-ES.acceptable> <http://snomed.info/id/190787008> "hipobetalipoproteinemia familiar homocigota"@es-es)
AnnotationAssertion(<http://snomed.info/field/es-ES.acceptable> <http://snomed.info/id/190787008> "síndrome de Bassen - Kornzweig"@es-es)
AnnotationAssertion(<http://snomed.info/field/es-ES.acceptable> <http://snomed.info/id/190787008> "síndrome de Bassen-Kornzweig"@es-es)
AnnotationAssertion(<http://snomed.info/field/es.acceptable> <http://snomed.info/id/190787008> "déficit de apolipoproteína B"@es)
AnnotationAssertion(<http://snomed.info/field/es.acceptable> <http://snomed.info/id/190787008> "enfermedad de Bassen-Kornzweig"@es)
AnnotationAssertion(<http://snomed.info/field/es.preferred> <http://snomed.info/id/190787008> "abetalipoproteinemia"@es)
AnnotationAssertion(<http://snomed.info/field/es.preferred> <http://snomed.info/id/190787008> "abetalipoproteinemia (trastorno)"@es)
AnnotationAssertion(rdfs:label <http://snomed.info/id/190787008> "Abetalipoproteinemia (disorder)")
```
When using owlapi in Java, I usually have the following functions:
```
private static void addAnnotation(OWLOntology ontModel, OWLEntity entity, OWLAnnotationProperty annotationProperty, String label, String lang) {
	OWLAxiom axiom = createOWLAnnotationAssertionAxiom(ontModel, entity, annotationProperty, label, lang);
	ontModel.getOWLOntologyManager().applyChange(new AddAxiom(ontModel, axiom));
}

private static OWLAxiom createOWLAnnotationAssertionAxiom(OWLOntology ontModel, OWLEntity entity, OWLAnnotationProperty annotationProperty, String label,
		String lang) {
	OWLAxiom owlAxiom = null;
	OWLDataFactory df = ontModel.getOWLOntologyManager().getOWLDataFactory();
	OWLAnnotation annotation = df.getOWLAnnotation(annotationProperty, df.getOWLLiteral(label, lang));
	owlAxiom = df.getOWLAnnotationAssertionAxiom(entity.getIRI(), annotation);
	return owlAxiom;
}
```
So you can use them as follows:
```
// get the ontology from its path
OWLOntology ontology = OWLManager.createOWLOntologyManager().loadOntologyFromOntologyDocument(new File(ontologyPath));

// get AnnotationProperty by its URI in the ontology (it would be created if it does not exist).
String annotationPropertyURI = "http://snomed.info/field/es-ES.acceptable";
OWLAnnotationProperty annotationProperty = ontology.getOWLOntologyManager().getOWLDataFactory().getOWLAnnotationProperty(annotationPropertyURI);

// Get the class we want to annotate
String owlClassURI = "http://snomed.info/id/190787008";
OWLClass owlClass = ontology.getOWLOntologyManager().getOWLDataFactory().getOWLClass(owlClassURI);

// Set the value of the annotation.
String value = "enfermedad de Bassen - Kornzweig";
String lang = "es-es";

// Add the annotation to the ontology.
addAnnotation(ontology, owlClass, annotationProperty, value, lang);
```
I agree with you, I'm going to close this issue because, in fact, it has been solved. I'll try to open a new ticket describing the functionality in the next days. Also I'll try to develop it on my own if I have enough time. I'll do a pull request if I achieve it.

Cheers and thanks again,
Francisco Abad.

**Comment by kaicode** [comment]
I'm reopening this issue to give others visibility of the ongoing feature development.
@mmorine has asked about similar functionality in #23, although with english terms.
@fanavarro just shout if I should take this on.

**Comment by fanavarro** [comment]
Hi @kaicode! I wanted to open a new post with this enhancement but I had have a lot of work the last weeks... So thanks for reopening this post, I think this functionality can be linked with this post. And of course, it would be great to have this feature to generate owl versions of Snomed!

**Comment by mmorine** [comment]
Great! Thanks for your reply. I'll keep an eye out for this feature since the synonyms and altLabels are necessary for my project.

@fanavarro can I confirm that you produced the below image using the java API rather than the command line tool? 

> The resulting ontology will show something like the following figure:
> ![imagen](https://user-images.githubusercontent.com/11376639/51168472-ac412180-18a9-11e9-8f05-405dc947190d.png)

**Comment by fanavarro** [comment]
Hi @mmorine! Indeed, I used the command line tool to generate the owl ontology, which only included rdfs:label, and then I enriched that ontology with the Spanish terms through the owlapi. I have the Spanish ministry version of Snomed CT, which includes SQL scripts to deploy a database with the contents of SNOMED. I developed a little java application that iterates over each concept in the SNOMED ontology and performs a query to retrieve the terms related to each concept. What I figured out is that each SNOMED extension provides two tables I can use to find the terms from a specific concept. The tabled named description table contains the terms, whereas the table named refset contains the information concerning their acceptability (if the term is preferred or alternative). My application generates SQL queries mixing these tables because it is possible that a term appears in description table of one extension and in the refset table of another extension.

I'm attaching my java application. Maybe it could be useful for someone. It is based on the SQL distribution provided by the Ministry of Spain and you will need an instance of the database, which is configured in the class called SnomedDBConfig. Also, the queries are generated in SQLQueryGenerator class and used in SnomedDAO class. SQLQueryGenerator has a list of refset and description tables, which are used later to generate the query; maybe these lists could be changed to obtain terms in other languages rather than Spanish. Also, I think that the SQL distribution of SNOMED is just a simple dump of the standard text files so that the files "sct2_description_spanishextensionsnapshot_es_int" and "der2_crefset_languagespanishextensionsnapshot_es_int" are just translated into two SQL tables with the same name and content. Maybe the SQL statements could be replaced by searches in the text files of the standard distribution.

[addSnomedTerms.zip](https://github.com/IHTSDO/snomed-owl-toolkit/files/2841812/addSnomedTerms.zip)
